core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java

```diff
@@ -706,4 +707,46 @@ public UserDto getUserOfBuddy(UUID userId, UUID buddyUserId)
 				.orElseThrow(() -> BuddyNotFoundException.notFoundForUser(userId, buddyUserId));
 		return UserDto.createInstanceWithBuddyData(userEntity, BuddyUserPrivateDataDto.createInstance(buddy));
 	}
+
+	@Transactional
+	public void processDeviceChange(UUID idOfRequestingUser, UUID relatedUserAnonymizedId, DeviceChange change,
+			UUID deviceAnonymizedId, Optional<String> oldName, Optional<String> newName)
+	{
+		User user = userService.getValidatedUserbyId(idOfRequestingUser);
+		Buddy buddy = user.getBuddyByUserAnonymizedId(relatedUserAnonymizedId);
+
+		switch (change)
+		{
+			case ADD:
+				addDeviceToBuddy(buddy, deviceAnonymizedId, newName.get());
+				break;
+			case RENAME:
+				updateDeviceNameForToBuddy(buddy, deviceAnonymizedId, newName.get());
+				break;
+			case DELETE:
+				removeDeviceFromBuddy(buddy, deviceAnonymizedId);
+				break;
+			default:
+				throw new IllegalArgumentException("Change " + change + " is unknown");
+		}
+	}
+
+	private void addDeviceToBuddy(Buddy buddy, UUID deviceAnonymizedId, String name)
+	{
+		BuddyDevice device = BuddyDevice.createInstance(name, deviceAnonymizedId, true); // Assume VPN is connected
```

Comment:
Couldn't this assumption lead to problems?

Reply:
Let's address this when handling the VPN status issue. See [YD-514](http://jira.yona.nu/browse/YD-514).